contracts-bedrock/contracts/L1/ResourceMetering.sol

```diff
@@ -0,0 +1,185 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.10;
+
+import { Math } from "@openzeppelin/contracts/utils/math/Math.sol";
+import { SignedMath } from "@openzeppelin/contracts/utils/math/SignedMath.sol";
+import { SlotPacking128x64x64 } from "../libraries/SlotPacking128x64x64.sol";
+import { Burn } from "../libraries/Burn.sol";
+
+/**
+ * @title ResourceMetering
+ * @notice ResourceMetering implements an EIP-1559 style resource metering system where pricing
+ * updates automatically based on current demand.
+ */
+contract ResourceMetering {
+    /**
+     * Along with the resource limit, determines the target resource limit.
+     */
+    int256 public constant ELASTICITY_MULTIPLIER = 2;
+
+    /**
+     * Denominator that determines max change on fee per block.
+     */
+    int256 public constant BASE_FEE_MAX_CHANGE_DENOMINATOR = 8;
+
+    /**
+     * Maximum amount of deposit gas that can be used within this block.
+     */
+    int256 public constant MAX_RESOURCE_LIMIT = 30000000;
```

Comment:
This value should be ~3-4 million. It is very important that this value is less than the L2 block gas limit. Assuming we start off at the same gas per second as ethereum, our blocks would be ~5million gas per block (30mil gas/block / 12s * 2s/block)

Reply:
We may want to consider the impact of a 5m gas limit, it's very possible that certain people expect large gas limits. I'll make sure to bring this up asap.

Reply:
One thing here is that this limit can be smaller than the L2 gas limit. One thing that is important is that we need to be careful on how we adjust the L2 gas limit - I don't believe that we can use the same mechanism as ethereum.

Reply:
Why is `MAX_RESOURCE_LIMIT` an `int` rather than a `uint`?

Reply:
@tynes it simplifies the code by avoiding the need to do lots of conversions between uint/int. We used ints because it simplified the code a bit by removing some unnecessary if statements.